File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int preidx=0; 
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder)
    {
        // 1. use the preidx to fill root, left , right 
        // 2. split inorder array based on the value inidx ->(0,inidx-1), (inidx+1, n);
        return genTree(preorder,inorder,0,inorder.size()-1);
    }

    int findIdx(vector<int>& inorder,int ele,int s,int e){
        for(int i=s;i<=e;i++){
            if(inorder[i]==ele){
                return i;
            }
        }
        return -1;
    }

    TreeNode* genTree(vector<int>& preorder,vector<int>& inorder,int l,int r)
    {
        if(l==r) {
            return new TreeNode(preorder[preidx++]);
        }

        if(l>r) return NULL;
        int node_val = preorder[preidx]; 
        
        //find the element in inorder and split nodes 
        int in_idx = findIdx(inorder,node_val,l,r);
        if(in_idx==-1) return NULL;
        preidx++;

        // 
        TreeNode* node = new TreeNode(inorder[in_idx]);
        node->left= genTree(preorder,inorder,l,in_idx-1);
        node->right = genTree(preorder,inorder,in_idx+1,r);
        return node;
    }
};

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self): 
        self.preidx = 0

    def findNode(self,inorder,l,r,target): 
        for i in range(l,r+1): 
            if inorder[i]==target:
                return i 
        
        return -1 

    def constructTree(self,preorder,inorder,l,r): 
        if l==r: 
            self.preidx+=1
            return TreeNode(inorder[l])
        if l>r: 
            return None
        inidx = self.findNode(inorder,l,r,preorder[self.preidx])
        if inidx==-1: return None 
        currnode = TreeNode(preorder[self.preidx])
        self.preidx+=1 
        currnode.left = self.constructTree(preorder,inorder,l,inidx-1) 
        currnode.right = self.constructTree(preorder,inorder,inidx+1,r)
        return currnode 

    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        m = len(inorder)-1
        return self.constructTree(preorder,inorder,0,m)
        
        